docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+```
+
+Now what if I want to have seperate MLP parameters for each item in `xs`?
+If this where "vanilla JAX" we could imagene writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
+But doing something like this in linen will actually fail:
+```
+class NaiveVmapMLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    mlps = MLP()
+    return jax.vmap(lambda mlp, x: mlp(x))(mlps, xs)  # fails
+```
+
+JAX will refuse to transform the `mlps` because its not a JAX Array or a simple container of arrays.
+We can't really blame JAX for refusing this dirty, underspecified task.
+After all it's not even clear what should happen here.
+The parameters inside MLP are not even initialized yet and we will need a seperate PRNG key for each group of parameters.
+
+We can fix this problem by first turning `MLP` into a pure init and apply function.
+Afterwards, we use the `param` method to store the parameters:
+```
+class ManualVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    mlp = MLP(parent=None)
+    init_fn = lambda rng, xs: jax.vmap(mlp.init, in_axes=0)(random.split(rng, xs.shape[0]), xs)['params']
+    apply_fn = jax.vmap(mlp.apply, in_axes=0)
+    mlp_params = self.param('mlp', init_fn, xs)
+    return apply_fn({'params': mlp_params}, xs)
+
+xs = jnp.ones((3, 4))
+variables = ManualVmap().init(random.PRNGKey(0), xs)
+print(jax.tree_map(jnp.shape, variables['params']))
+"""==>
+{
+    mlp: {
+        hidden: {
+            bias: (3, 4),
+            kernel: (3, 4, 4),
+        },
+        out: {
+            bias: (3, 1),
+            kernel: (3, 4, 1),
+        },
+    },
+}
+"""
+```
+
+Here `MLP(parent=None)` creates a detached instance of `MLP`.
+This avoids reserving a name for the submodule inside the current module.
+Although not strictly necessary, this also ensures we cannot accidentaly use the MLP instance in a stateful way and we are forced to use it through either `.init` or `.apply`.
+
+This example is still relatively concise but it already takes a few extra "bookkeeping" statements to make it work.
+However, this implementation has a number of limitations:
+1. During initialization we call the submodule twice through `init_fn` & `apply_fn`. If the submodule used the same trick to do
+   functional tranformation we will end up executing a lot of code as the number of module calls grows like 2^d where d is the number of
+   nested function transformations.
+2. The implemenation assumes the submodule only requires the parameter RNG sequence.
+3. The implementation assumes we init parameters once and we don't create any state variables either during init or apply.
+
+Point 3 in particular makes manual functionalization cumbersome. Feel free to try and extend the above example with a `nn.BatchNorm`
+layer in the `MLP` module. This should cause a number of complexites further growing the implementation. Like storing the updated
```

Comment:
Suggestion (and correct me if I misunderstood this)

```suggestion
layer in the `MLP` module. This should increase the complexity of the implementation. Like storing the updated
```

(the original didn't look right (grammar) IMO)